File: source/C++/ValidAnagram.cpp
// https://leetcode.com/problems/valid-anagram
#include <string>
#include <unordered_map>

class Solution
{
public:
    bool isAnagram(std::string s, std::string t)
    {
        if (s.size() != t.size()) return false;

        std::unordered_map<char, int> letters;

        for (const char& c : s) ++letters[c];
        for (const char& c : t) --letters[c];

        for (const auto& [key, value] : letters)
        {
            if (value != 0) return false;
        }

        return true;
    }
};

File: CSharpPractice/source/ValidAnagram.cs
// https://leetcode.com/problems/valid-anagram
using System.Collections.Generic;

namespace LeetCodePractice.ValidAnagram
{
    public class Solution
    {
        public bool IsAnagram(string s, string t)
        {
            if (s.Length != t.Length) return false;

            Dictionary<char, int> letters = new Dictionary<char, int>();

            foreach (char c in s)
            {
                if (letters.ContainsKey(c))
                {
                    ++letters[c];
                }
                else
                {
                    letters[c] = 1;
                }
            }
                
            foreach (char c in t)
            {
                if (letters.ContainsKey(c))
                {
                    --letters[c];
                }
                else
                {
                    return false; // character didnt appear in first string
                }
            }

            foreach (int value in letters.Values)
            {
                if (value != 0) return false;
            }

            return true;
        }
    }
}